### PR TITLE
Convert `aten.ceil` to `ttnn.ceil`

### LINKED ISSUE #493: aten.ceil.default

### PR DESCRIPTION
**Pull request by jdh8**
### Ticket
- Working around tenstorrent/tt-metal#12671
- Resolves #493

### Problem description
Convert `aten.ceil` to `ttnn.ceil` and probably other rounding ops

Currently, 1-D cases fail because `ttnn.ceil` produces 2-D results.  For instance, `ttnn.ceil` takes a (1066,) tensor but produces a (1, 1066) tensor.

```
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape0] - AssertionError: list(expected_pytorch_result.shape)=[1066] vs list(actual_pytorch_result.shape)=[1, 1066]
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape1] - AssertionError: list(expected_pytorch_result.shape)=[120] vs list(actual_pytorch_result.shape)=[1, 120]
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape2] - AssertionError: list(expected_pytorch_result.shape)=[128] vs list(actual_pytorch_result.shape)=[1, 128]
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape3] - AssertionError: list(expected_pytorch_result.shape)=[160] vs list(actual_pytorch_result.shape)=[1, 160]
```

I doubt if this happens to other elementwise ops, too.

### What's changed
- [x] Convert `aten.ceil` to `ttnn.ceil`
- [x] Test conversion of `aten.ceil`
- [x] Convert `aten.round` to `ttnn.round`
- [x] Test conversion of `aten.round`
- [x] Convert `aten.trunc` to `ttnn.trunc`
- [x] Test conversion of `aten.trunc`

### CODE REVIEW

**Comment by ayerofieiev-tt** [comment]
@jdh8 , lets fire a ticket and lets overcome in compiler by squeezing the dimension?

**Comment by jdh8** [comment]
I'll probably deal with this PR after #113 and #170.  I suggest patching once for all univariate functions because there are many of them.

**Comment by jdh8** [comment]
Unfortunately, the workaround by squeezing out the extraneous dimension for 1-D tensors (e6e9ef8e556bc60152a5de5fc6373b015b65a464) still leaves errors.  It cannot resolve tenstorrent/tt-metal#12671.
```
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape0] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape2] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape3] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_cos.py::test_cos[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_erf.py::test_erf[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_exp.py::test_exp[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_floor.py::test_floor[input_shape4] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_gelu.py::test_gelu[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_neg.py::test_neg[input_shapes0] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape0-7-True] - AssertionError: assert 0 == 1
FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape1-3-True] - AssertionError: assert 0 == 1
FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape2-5-True] - AssertionError: assert 0 == 1
FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape3-1-True] - AssertionError: assert 0 == 1
FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape4-1-True] - AssertionError: assert 0 == 1
FAILED tests/lowering/eltwise/unary/test_rsqrt.py::test_rsqrt[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
FAILED tests/lowering/eltwise/unary/test_sqrt.py::test_sqrt[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
========================================================================= 17 failed, 43 passed in 39.79s ==========================================================================
```

**Comment by jerrysky3** [comment]
> Unfortunately, the workaround by squeezing out the extraneous dimension for 1-D tensors ([e6e9ef8](https://github.com/tenstorrent/pytorch2.0_ttnn/commit/e6e9ef8e556bc60152a5de5fc6373b015b65a464)) still leaves errors. It cannot resolve [tenstorrent/tt-metal#12671](https://github.com/tenstorrent/tt-metal/issues/12671).
> 
> ```
> FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape0] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape2] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_ceil.py::test_ceil[input_shape3] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_cos.py::test_cos[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_erf.py::test_erf[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_exp.py::test_exp[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_floor.py::test_floor[input_shape4] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_gelu.py::test_gelu[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_neg.py::test_neg[input_shapes0] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape0-7-True] - AssertionError: assert 0 == 1
> FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape1-3-True] - AssertionError: assert 0 == 1
> FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape2-5-True] - AssertionError: assert 0 == 1
> FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape3-1-True] - AssertionError: assert 0 == 1
> FAILED tests/lowering/eltwise/unary/test_remainder.py::test_remainder[input_shape4-1-True] - AssertionError: assert 0 == 1
> FAILED tests/lowering/eltwise/unary/test_rsqrt.py::test_rsqrt[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> FAILED tests/lowering/eltwise/unary/test_sqrt.py::test_sqrt[input_shape1] - RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:170: normalized_index >= 0 and normalized_index < rank
> ========================================================================= 17 failed, 43 passed in 39.79s ==========================================================================
> ```

It seems like the problem is the tensor returned by `ttnn.ceil` is in TILE_LAYOUT, which `ttnn.squeeze` don't know how to turn an 2D tiled tensor into 1d. It works for me if the tensor is first converted into ROW_MARJO_LAYOUT:
```
result_after = ttnn.ceil(input_tensor)
result_after = ttnn.to_layout(result_after, ttnn.ROW_MAJOR_LAYOUT)
result_after = ttnn.squeeze(result_after, 0)
```

**Comment by jdh8** [comment]
My test results:
```
(python_env) jdh8@tt-loudbox:~/pytorch2.0_ttnn$ pytest tests/lowering/eltwise/unary/test_round.py 
=============================== test session starts ================================
platform linux -- Python 3.8.10, pytest-7.2.2, pluggy-1.5.0
rootdir: /home/jdh8/pytorch2.0_ttnn/tests, configfile: pytest.ini
plugins: split-0.8.2, anyio-4.5.2, xdist-3.6.1, dash-2.15.0, timeout-2.2.0
collected 5 items                                                                  

tests/lowering/eltwise/unary/test_round.py .....                             [100%]

================================ 5 passed in 3.40s =================================
                 Device | INFO     | Closing user mode device drivers
```

**Comment by jdh8** [comment]
All the failing ops are `aten.max_pool2d_with_indices_backward.default` from U-Net and friends.  We can catch these ops with

```sh
pytest tests/autogen_op/U*-train/*aten_max_pool2d_with_indices_backward_default.py
```

**Comment by jdh8** [comment]
Workaround issues left:
- tenstorrent/tt-metal#12671
- #508

**Comment by jdh8** [comment]
Tests are running at https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/12038678310

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 1]
```diff
@@ -470,6 +402,25 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+                    result = g.call_function(ttnn.to_layout, (result, TtnnRowMajorLayout()))
+                    result = g.call_function(ttnn.squeeze, (result, 0))
+
+                return result
+
+            if node.target in TTNN_POINTWISE_UNARY_OPS:
+                return unsqueeze_to_2d(TTNN_POINTWISE_UNARY_OPS[node.target])
+
+            if node.target == torch.ops.aten.round.default:
+                return unsqueeze_to_2d(ttnn.round, (args[0],), {"decimals": 0})
```
Comment (ayerofieiev-tt) [inline torch_ttnn/passes/lowering/to_tt_pass.py:422]: Must update ttnn op binding to default the argument to 0?

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 2]
```diff
@@ -470,6 +402,25 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+                result = g.call_function(code, args, kwargs)
+
+                if ndims == 1:
+                    result = g.call_function(ttnn.to_layout, (result, TtnnRowMajorLayout()))
+                    result = g.call_function(ttnn.squeeze, (result, 0))
+
+                return result
+
+            if node.target in TTNN_POINTWISE_UNARY_OPS:
+                return unsqueeze_to_2d(TTNN_POINTWISE_UNARY_OPS[node.target])
```
Comment (ayerofieiev-tt) [inline torch_ttnn/passes/lowering/to_tt_pass.py:419]: What do we unsqueeze here? Why only for unary?

Comment (jdh8) [reply torch_ttnn/passes/lowering/to_tt_pass.py]: Thanks for inspiring me with a simpler algorithm.  Reshape the result back when `ndims < 2`.

File: torch_ttnn/passes/lowering/to_tt_pass.py (hunk 1 above)
Comment (jdh8) [reply torch_ttnn/passes/lowering/to_tt_pass.py]: That would be great!  I once tried at tenstorrent/tt-metal#13851 but in vain.

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 3]
```diff
@@ -471,6 +403,20 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
                 output = batch_norm_post_process(output, shape, weight, bias)
                 return g.call_function(target_wrappers.pack_to_tuple, (output,))
 
+            # NOTE(jdh8): Workaround for tenstorrent/tt-metal#12671
+            # Passing a tensor shaped `(N,)` to the kernel results in `(1, N)`.
+            # Reshape the tensor back to get the correct shape.
+            def reshape_1d(code, args=args, kwargs=kwargs):
+                shape = node.meta["val"].size()
+                result = g.call_function(code, args, kwargs)
+                return result if len(shape) > 1 else g.call_function(ttnn.reshape, (result, shape))
```
Comment (jdh8) [inline torch_ttnn/passes/lowering/to_tt_pass.py:412]: This new implementation only introduces 0&ndash;1 op.  I also rename this tool function to reflect what it does.

**Comment by jdh8** [comment]
I cannot locally reproduce the errors found in CI.
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/12073595035/job/33670553655#step:5:26198
```
==== 4 passed, 156 deselected, 2 xfailed, 8 warnings in 1076.48s (0:17:56) =====
|                 path                 |     passed     |   xfailed    | subtotal |
| ------------------------------------ | -------------- | ------------ | -------: |
| models/falcon/test_falcon.py         | :green_circle: |              |        1 |
| models/flan_t5/test_flan_t5.py       | :green_circle: | :red_circle: |        2 |
| models/glpn_kitti/test_glpn_kitti.py | :green_circle: | :red_circle: |        2 |
| models/gpt2/test_gpt2.py             | :green_circle: |              |        1 |
| TOTAL                                |              4 |            2 |        6 |
```

However, ResNet raises these errors only locally.
```
ERROR tests/models/resnet/test_resnet.py::test_resnet[train] - TypeError: ResNet18-train compiled failed to run.
ERROR tests/models/resnet/test_resnet.py::test_resnet[eval] - TypeError: ResNet18 compiled failed to run.
```

**Comment by ayerofieiev-tt** [comment]
@kevinwuTT if you can suggest anything to enable debugging of this case

**Comment by jdh8** [comment]
Tests finally pass!
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/12420284637

File: torch_ttnn/passes/lowering/to_tt_pass.py (hunk 3 above)
Comment (ayerofieiev-tt) [reply torch_ttnn/passes/lowering/to_tt_pass.py:427]: should this be a call to ttnn.unsqueeze?

**Review by ayerofieiev-tt — APPROVED** [review]

**Comment by ayerofieiev-tt** [comment]
Tests pass https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/12420284637

**Pull request merged**